pkg/utils/errorutils.go

```diff
@@ -0,0 +1,180 @@
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package utils
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const MAX_SIMILAR_ERRORS_TO_LOG = 5 // Maximum number of similar errors to store
+
+var qidToBatchErrorMap map[uint64]*BatchError
+var qidToBatchErrorMapLock *sync.RWMutex
+
+// BatchErrorData holds error information for a specific error key
+type BatchErrorData struct {
+	errors  []error
+	counter uint32
+}
+
+// BatchError manages batched errors with thread safety
+type BatchError struct {
+	beMap *sync.Map
+	qid   int64
+}
+
+// ErrorWithCode combines an error with a standardized error code
+type ErrorWithCode struct {
+	code string
+	err  error
+}
+
+func init() {
+	qidToBatchErrorMap = make(map[uint64]*BatchError)
+	qidToBatchErrorMapLock = &sync.RWMutex{}
+}
+
+// NewBatchError creates a new BatchError instance
+func NewBatchError() *BatchError {
+	return &BatchError{
+		beMap: &sync.Map{},
+		qid:   -1,
+	}
+}
+
+// AddError adds an error to the batch with thread safety using sync.Map
+func (be *BatchError) AddError(errKey string, err error) {
+	if err == nil {
+		return
+	}
+
+	if errorWithCode, ok := err.(*ErrorWithCode); ok {
+		errKey = fmt.Sprintf("%s:%s", errKey, errorWithCode.code)
+	}
+
+	// Get or create BatchErrorData atomically
+	value, _ := be.beMap.LoadOrStore(errKey, &BatchErrorData{
+		errors: make([]error, 0, MAX_SIMILAR_ERRORS_TO_LOG),
+	})
+	data := value.(*BatchErrorData)
+
+	atomic.AddUint32(&data.counter, 1)
+
+	// Only store error if we haven't reached the limit
+	if len(data.errors) < MAX_SIMILAR_ERRORS_TO_LOG {
+		data.errors = append(data.errors, err)
+	}
+}
+
+// LogAllErrors logs all accumulated errors
+// Send qid as -1 if qid is not available
+func (be *BatchError) LogAllErrors() {
+	qidMsg := ""
+	if be.qid >= 0 {
+		qidMsg = fmt.Sprintf("qid=%v, ", be.qid)
+	}
+
+	be.beMap.Range(func(key, value interface{}) bool {
+		data := value.(*BatchErrorData)
+		log.WithFields(log.Fields{
+			"count":   data.counter,
+			"samples": data.errors,
+		}).Errorf("%s ErrorKey=%v", qidMsg, key)
+		return true
+	})
+}
+
+// Reset clears all stored errors
+func (be *BatchError) Reset() {
+	be.beMap = &sync.Map{}
+}
+
+// HasErrors returns true if there are any stored errors
+func (be *BatchError) HasErrors() bool {
```

Comment:
nit: you chose a good name for this function, so you don't really need this comment either. But you can keep it if you want